corehq/ex-submodules/casexml/apps/phone/restore.py

```diff
@@ -132,27 +151,39 @@ def __add__(self, other):
 
         return response
 
-    def compose(self):
+    def finalize(self):
         """
         Creates the final file with start and ending tag
         """
-        # Add 1 to num_items to account for message element
-        items = self.items_template.format(self.num_items + 1) if self.items else ''
-        self.response.write(self.start_tag_template.format(
-            items=items,
-            username=self.username,
-            nature=ResponseNature.OTA_RESTORE_SUCCESS
-        ))
+        with open(self.get_filename(), 'w') as response:
+            # Add 1 to num_items to account for message element
+            items = self.items_template.format(self.num_items + 1) if self.items else ''
+            response.write(self.start_tag_template.format(
+                items=items,
+                username=self.username,
+                nature=ResponseNature.OTA_RESTORE_SUCCESS
+            ))
 
-        self.response_body.seek(0)
-        shutil.copyfileobj(self.response_body, self.response)
+            self.response_body.seek(0)
+            shutil.copyfileobj(self.response_body, response)
 
-        self.response.write(self.closing_tag)
-        return self.get_filename()
+            response.write(self.closing_tag)
+        
+        self.finalized = True
+        self.close()
 
-    def close(self):
-        self.response_body.close()
-        self.response.close()
+    def get_cache_payload(self, full=False):
+        return {
+            'is_file': True,
+            'data': self.get_filename() if not full else open(self.get_filename(), 'r')
```

Comment:
oh oops i missed the part where this didn't read the file and just returned the handle. agreed we shouldn't read the whole file into memory